### PR TITLE
ci: upload nvimlog as artifact

### LINKED ISSUE #18373: build: write minimal logs to stdout, full log as CI artifact
## Problem

Build logs in CI are mostly noise. But sometimes we need the full details of the log.

https://github.com/neovim/neovim/pull/17906#issuecomment-1109171265

## Solution

1. Modify `test/busted/outputHandlers/nvim.lua` so that:
    1. when running in CI, it writes the full logs (as today) to a known location
        1. make the full log available as a GHA artifact.
    1. when running in CI, it writes _minimal_ logs to stdout
        1. omits _successful_ tests
        1. prints a dot (.) every few tests (to avoid "hang" detection in CI)
1. Modify `src/nvim/testdir/runnvim.sh` in a similar way (or pipe it through the same filter as `test/busted/outputHandlers/nvim.lua`)

---

## See also

- https://github.com/neovim/neovim/pull/11009 will reduce the "server failed to start" spam in the logs
- https://github.com/neovim/neovim/pull/8519

### PR DESCRIPTION
**Pull request by dundargoc**
This keeps the CI logs relatively clean while still giving the
opportunity aqcuire the nvimlog information if needed.

Closes https://github.com/neovim/neovim/issues/18373

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -214,6 +213,12 @@ jobs:
         name: Show logs
         run: cat $(find "$LOG_DIR" -type f)
 
+      - if: success() || failure() && steps.abort_job.outputs.status == 'success'
+        uses: actions/upload-artifact@v3
+        with:
+          name: nvimlog-${{ join(matrix.*, '-') }}
+          path: ${{ env.NVIM_LOG_FILE }}
+
```
Comment (zeertzjq) [inline .github/workflows/test.yml:221]: Default retention period may be a bit too long

**Review by justinmk — CHANGES_REQUESTED** [review]
👍 Making the `NVIM_LOG_FILE` available as an artifact is useful, no objection there.

But https://github.com/neovim/neovim/issues/18373 (which is up for debate) is about the build/test output, not `NVIM_LOG_FILE`. 

Showing (the tail of) `NVIM_LOG_FILE` is not a problem, imo. Hiding it doesn't fix anything, it just makes the problem invisible.

- If there are truly _noise_ errors being shown, those should be explicitly fixed in the tests that cause them.
- The libuv "ref" warnings also need to be fixed.
- Any other warnings should be fixed or explicitly ignored in the relevant test.

**Comment by dundargoc** [comment]
> If there are truly noise errors being shown, those should be explicitly fixed in the tests that cause them.

This sentiment doesn't mean anything because we have a fundamental disagreement on what constitutes noise.

**Comment by justinmk** [comment]
Example?

If the tests are printing things that aren't telling us anything actionable, then that's noise.

There are several actions that can be taken for each case:

- change the log-level of the message (CI does only prints WARN/ERROR logs)
- fix the test so that it logs to a temporary NVIM_LOG_FILE instead of the shared one
- fix the cause of the error/warning

Where is the fundamental disagreement?

**Comment by dundargoc** [comment]
I just have absolutely no idea what I'm supposed to do with

```
ERR 2023-09-18T12:29:38.911 T1327.12039.0 libuv_process_spawn:97: uv_spawn(/usr/bin/pwd) failed: permission denied
ERR 2023-09-18T12:29:46.525 T1368.12291.0 libuv_process_spawn:97: uv_spawn(nosuchshell) failed: no such file or directory
ERR 2023-09-18T12:29:46.530 T1368.12291.0 libuv_process_spawn:97: uv_spawn(nosuchshell) failed: no such file or directory
ERR 2023-09-18T12:29:54.388 T1425.12699.0/c tui_mode_change:1142: uv_tty_set_mode failed: i/o error
ERR 2023-09-18T12:29:54.511 T1426.12712.0/c os_set_cloexec:475: Failed to get flags on descriptor 3: Bad file descriptor
ERR 2023-09-18T12:34:28.134 T3912.19804.0 loop_close:174: uv_loop_close() hang?
[--I] signal   0x5615173c25a8
[-AI] async    0x5615173c23f0
[RA-] timer    0x610000003a40
ERR 2023-09-18T12:34:37.902 T3918.19832.0 loop_close:174: uv_loop_close() hang?
[--I] signal   0x561cf56e75a8
[-AI] async    0x561cf56e73f0
[RA-] fs_poll  0x60b000000460
[-AI] timer    0x6180000000a8
ERR 2023-09-18T12:36:30.159 nvim.24339.0 tui_mode_change:1146: uv_tty_set_mode failed: inappropriate ioctl for device
ERR 2023-09-18T12:36:50.788 T4881.25003.0 decor_provider_error:32: error in provider ns1.start: Error executing lua: [string "<nvim>"]:4: Foo
stack traceback:
	[C]: in function 'error'
	[string "<nvim>"]:4: in function <[string "<nvim>"]:3>
ERR 2023-09-18T12:37:05.252 T5011.25446.0 os_set_cloexec:475: Failed to get flags on descriptor 3: Bad file descriptor
ERR 2023-09-18T12:39:36.863 T6535.30518.0 libuv_process_spawn:97: uv_spawn(fake_shell) failed: no such file or directory
```

What does any of this mean? Do our non-veteran contributors have any chance of possible understanding what this means?

Let's take the first error. It seems `/usr/bin/pwd` failed because of permission problems? Except the failing test never runs `pwd`. I guess one function in the test called a function in luv, which in turn called `pwd`? How do I figure out which function? Am I supposed to manually trace every function call until I find `pwd` in libuv? What's the plan here?

These logs aren't helpful to me, and I'm willing to bet they aren't for the majority of contributors (including core contributors).

**Comment by jamessan** [comment]
Not sure what test run you're looking at, so I picked the [ASAN](https://github.com/neovim/neovim/actions/runs/6207337325/job/16853574984?pr=25162) job for this PR.

> Let's take the first error. It seems `/usr/bin/pwd` failed because of permission problems? Except the failing test never runs `pwd`.

Ok, that's this error:

```
ERR 2023-09-18T12:29:38.911 T1327.12039.0 libuv_process_spawn:97: uv_spawn(/usr/bin/pwd) failed: permission denied
```

The `T1327` bit there is the test that was running, which is ["jobs error on non-executable `cwd`"](https://github.com/neovim/neovim/actions/runs/6207337325/job/16853574984?pr=25162#step:12:1878).

That [test](https://github.com/neovim/neovim/blob/46402c16c0c38701469c52fb28d16f2483cc7a72/test/functional/core/job_spec.lua#L227-L236) does try to run `pwd` and expects it to fail with permission denied.

This sound like the "fix the test so that it logs to a temporary NVIM_LOG_FILE instead of the shared one" scenario.

Associating the logs to the test used to be a lot more problematic before the "T####" thing was added.

**Comment by dundargoc** [comment]
> The T1327 bit there is the test that was running

...literally how are you supposed to deduce that? And no, I don't really think it means much to claim it's obvious if you wrote it/were present when it was written. Is there a technical limitation to why we can't add the test name to the logs at least?

**Comment by jamessan** [comment]
> > The T1327 bit there is the test that was running
> 
> ...literally how are you supposed to deduce that?

By reading https://github.com/neovim/neovim/blob/master/test/README.md#debugging-tests

Although, that should clarify what the other numbers are after the T### part. I believe the second number is the PID, not sure off-hand about the third.

**Comment by dundargoc** [comment]
Meh, I missed that the README had been updated. I still don't understand why we can't print the test name in the logs?

**Comment by dundargoc** [comment]
What specifically are you referring to when you say "temporary NVIM_LOG_FILE"? Is it a throwaway log file that we don't use, something like a cross-platform `/dev/null` or are you thinking we should do something with the file?

**Pull request closed without merging**